src/Sentry/GlobalSessionManager.cs

```diff
@@ -0,0 +1,240 @@
+            // Note: in the future, this probably has to be synchronized across multiple processes too.
+            lock (_lock)
+            {
+                // We may have acquired the lock after another thread has already resolved
+                // installation ID, so check the cache one more time before proceeding with I/O.
+                if (!string.IsNullOrWhiteSpace(_cachedInstallationId))
+                {
+                    return _cachedInstallationId;
+                }
+
+                var id =
+                    TryGetPersistentInstallationId() ??
+                    TryGetHardwareInstallationId();
+
+                if (!string.IsNullOrWhiteSpace(id))
+                {
+                    _options.DiagnosticLogger?.LogDebug(
+                        "Resolved installation ID '{0}'.",
+                        id
+                    );
+                }
+                else
+                {
+                    _options.DiagnosticLogger?.LogDebug(
+                        "Failed to resolve installation ID."
+                    );
+                }
+
+                return _cachedInstallationId = id;
+            }
+        }
+
+        public SessionUpdate? StartSession()
+        {
+            // Extract release
+            var release = ReleaseLocator.Resolve(_options);
+            if (string.IsNullOrWhiteSpace(release))
+            {
+                // Release health without release is just health (useless)
+                _options.DiagnosticLogger?.LogError(
+                    "Failed to start a session because there is no release information."
+                );
+
+                return null;
+            }
+
+            // Extract other parameters
+            var environment = EnvironmentLocator.Resolve(_options);
+            var distinctId = TryGetInstallationId();
+
+            // Create new session
+            var session = new Session(distinctId, release, environment);
+
+            // Set new session and check whether we ended up overwriting an active one in the process
+            var previousSession = Interlocked.Exchange(ref _currentSession, session);
+            if (previousSession is not null)
+            {
+                _options.DiagnosticLogger?.LogWarning(
+                    "Starting a new session while an existing one is still active."
+                );
+
+                // End previous session
+                EndSession(previousSession, SessionEndStatus.Exited);
+            }
+
+            _options.DiagnosticLogger?.LogInfo(
+                "Started new session (SID: {0}; DID: {1}).",
+                session.Id, session.DistinctId
+            );
+
+            return session.CreateUpdate(true);
+        }
+
+        public SessionUpdate? ReportError()
+        {
+            if (_currentSession is { } session)
+            {
+                session.ReportError();
```

Comment:
Technically this could be an 'ended' session since another thread could have raced and closed/also called reported error on it. We either need to be able to flip the `errors` atomically or we need to sync all methods here to have a consistent view of things.

Considering worst case scenario is that we double report a session with the same state (errored+closed or errored+2) this isn't critical so we can merge as-is.